aws/rust-runtime/aws-types/src/profile/parse.rs

```diff
@@ -0,0 +1,333 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+//! Profile file parsing
+//!
+//! This file implements profile file parsing at a very literal level. Prior to actually being used,
+//! profiles must be normalized into a canonical form. Constructions that will eventually be
+//! deemed invalid are accepted during parsing such as:
+//! - keys that are invalid identifiers: `a b = c`
+//! - profiles with invalid names
+//! - profile name normalization (`profile foo` => `foo`)
+
+use crate::profile::source::File;
+use std::borrow::Cow;
+use std::collections::HashMap;
+use std::fmt::{self, Display, Formatter};
+
+/// A set of profiles that still carries a reference to the underlying data
+pub type RawProfileSet<'a> = HashMap<&'a str, HashMap<&'a str, Cow<'a, str>>>;
+
+/// Characters considered to be whitespace by the spec
+///
+/// Profile parsing is actually quite strict about what is and is not whitespace, so use this instead
+/// of `.is_whitespace()` / `.trim()`
+const WHITESPACE: &[char] = &[' ', '\t'];
+const COMMENT: &[char] = &['#', ';'];
+
+/// Location for use during error reporting
+#[derive(Clone, Debug, Eq, PartialEq)]
+struct Location {
+    line_number: usize,
+    path: String,
+}
+
+/// An error encountered while parsing a profile
+pub struct ProfileParseError {
+    /// Location where this error occurred
+    location: Location,
+
+    /// Error message
+    message: String,
+}
+
+impl Display for ProfileParseError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "error parsing {} on line {}:\n  {}",
+            self.location.path, self.location.line_number, self.message
+        )
+    }
+}
+
+/// Validate that a line represents a valid subproperty
+///
+/// - Subproperties looks like regular properties (`k=v`) that are nested within an existing property.
+/// - Subproperties must be validated for compatibility with other SDKs, but they are not actually
+/// parsed into structured data.
+fn validate_subproperty(value: &str, location: Location) -> Result<(), ProfileParseError> {
+    if value.trim_matches(WHITESPACE).is_empty() {
+        Ok(())
+    } else {
+        parse_property_line(value)
+            .map_err(|err| err.into_error("sub-property", location))
+            .map(|_| ())
+    }
+}
+
+fn is_empty_line(line: &str) -> bool {
+    line.trim_matches(WHITESPACE).is_empty()
+}
+
+fn is_comment_line(line: &str) -> bool {
+    line.starts_with(COMMENT)
```

Comment:
Is a line with some whitespace before the comment character not considered a comment? I.e.,
```ini
[my-profile]
   # indenting my comment and properties
   something = something
```

I don't know if that's valid or not.

Are comments on the same line as a property/profile name not allowed?

Edit: Looks like you handled this in `prepare_line`, and that this is merely a short-circuit. Although, maybe the short-circuit should also work on comment lines that are prefixed with whitespace.

Reply:
comments can't be prefixed with whitespace (because a non-empty line has meaning). The format dictates that a line that starts with whitespace is actually a subproperty